### PR TITLE
Add gesture classification for demo

### PR DESCRIPTION
**Pull request by yqxzy**
Partially resolves https://github.com/opencv/opencv_zoo/issues/127 

Add handpose estimation (0-9)
image

### CODE REVIEW

File: models/handpose_estimation_mediapipe/demo.py [hunk 1]
```diff
@@ -27,16 +31,21 @@ def str2bool(v):
     help_msg_backends += "; {:d}: TIMVX"
     help_msg_targets += "; {:d}: NPU"
 except:
-    print('This version of OpenCV does not support TIM-VX and NPU. Visit https://github.com/opencv/opencv/wiki/TIM-VX-Backend-For-Running-OpenCV-On-NPU for more information.')
+    print(
+        'This version of OpenCV does not support TIM-VX and NPU. Visit https://github.com/opencv/opencv/wiki/TIM-VX-Backend-For-Running-OpenCV-On-NPU for more information.')
```
Comment (WanliZhong) [inline models/handpose_estimation_mediapipe/demo.py:35]: unnecessary modification

File: models/handpose_estimation_mediapipe/demo.py [hunk 2]
```diff
@@ -27,16 +31,21 @@ def str2bool(v):
...
 
 parser = argparse.ArgumentParser(description='Hand Pose Estimation from MediaPipe')
 parser.add_argument('--input', '-i', type=str, help='Path to the input image. Omit for using default camera.')
-parser.add_argument('--model', '-m', type=str, default='./handpose_estimation_mediapipe_2023feb.onnx', help='Path to the model.')
+parser.add_argument('--model', '-m', type=str, default='./handpose_estimation_mediapipe_2023feb.onnx',
+                    help='Path to the model.')
```
Comment (WanliZhong) [inline models/handpose_estimation_mediapipe/demo.py:40]: unnecessary modification

File: models/handpose_estimation_mediapipe/demo.py [hunk 3]
```diff
@@ -27,16 +31,21 @@ def str2bool(v):
...
 parser.add_argument('--target', '-t', type=int, default=targets[0], help=help_msg_targets.format(*targets))
-parser.add_argument('--conf_threshold', type=float, default=0.9, help='Filter out hands of confidence < conf_threshold.')
-parser.add_argument('--save', '-s', type=str, default=False, help='Set true to save results. This flag is invalid when using camera.')
-parser.add_argument('--vis', '-v', type=str2bool, default=True, help='Set true to open a window for result visualization. This flag is invalid when using camera.')
+parser.add_argument('--conf_threshold', type=float, default=0.9,
+                    help='Filter out hands of confidence < conf_threshold.')
+parser.add_argument('--save', '-s', type=str, default=False,
+                    help='Set true to save results. This flag is invalid when using camera.')
+parser.add_argument('--vis', '-v', type=str2bool, default=True,
+                    help='Set true to open a window for result visualization. This flag is invalid when using camera.')
```
Comment (WanliZhong) [inline models/handpose_estimation_mediapipe/demo.py:48]: unnecessary modification

File: models/handpose_estimation_mediapipe/demo.py [hunk 4]
```diff
@@ -108,7 +117,8 @@ def draw_lines(image, landmarks, is_draw_point=True, thickness=2):
         # draw box
         cv.rectangle(display_screen, (bbox[0], bbox[1]), (bbox[2], bbox[3]), (0, 255, 0), 2)
         # draw handedness
-        cv.putText(display_screen, '{}'.format(handedness_text), (bbox[0], bbox[1] + 12), cv.FONT_HERSHEY_DUPLEX, 0.5, (0, 0, 255))
+        cv.putText(display_screen, '{}'.format(handedness_text), (bbox[0], bbox[1] + 12), cv.FONT_HERSHEY_DUPLEX, 0.5,
+                   (0, 0, 255))
```
Comment (WanliZhong) [inline models/handpose_estimation_mediapipe/demo.py:121]: unnecessary modification

File: models/handpose_estimation_mediapipe/demo.py [hunk 5]
```diff
@@ -146,6 +156,126 @@ def draw_lines(image, landmarks, is_draw_point=True, thickness=2):
...
+    except:
+        angle_ = 65535.
+    if angle_ > 180.:
+        angle_ = 65535.
+    return angle_
+
+
+def hand_angle(hand_):
+    angle_list = []
+    # ---------------------------- thumb 大拇指角度
```
Comment (WanliZhong) [inline models/handpose_estimation_mediapipe/demo.py:176]: Don't leave Chinese comments

File: models/handpose_estimation_mediapipe/demo.py [hunk 6]
```diff
@@ -146,6 +156,126 @@ def draw_lines(image, landmarks, is_draw_point=True, thickness=2):
...
+    angle_ = vector_2d_angle(
+        ((int(hand_[0][0]) - int(hand_[2][0])), (int(hand_[0][1]) - int(hand_[2][1]))),
+        ((int(hand_[3][0]) - int(hand_[4][0])), (int(hand_[3][1]) - int(hand_[4][1])))
+    )
+    angle_list.append(angle_)
+    # ---------------------------- index 食指角度
```
Comment (WanliZhong) [inline models/handpose_estimation_mediapipe/demo.py:182]: Don't leave Chinese comments

File: models/handpose_estimation_mediapipe/demo.py [hunk 7]
```diff
@@ -146,6 +156,126 @@ def draw_lines(image, landmarks, is_draw_point=True, thickness=2):
...
+    angle_ = vector_2d_angle(
+        ((int(hand_[0][0]) - int(hand_[6][0])), (int(hand_[0][1]) - int(hand_[6][1]))),
+        ((int(hand_[7][0]) - int(hand_[8][0])), (int(hand_[7][1]) - int(hand_[8][1])))
+    )
+    angle_list.append(angle_)
+    # ---------------------------- middle 中指角度
```
Comment (WanliZhong) [inline models/handpose_estimation_mediapipe/demo.py:188]: Don't leave Chinese comments

File: models/handpose_estimation_mediapipe/demo.py [hunk 8]
```diff
@@ -146,6 +156,126 @@ def draw_lines(image, landmarks, is_draw_point=True, thickness=2):
...
+    angle_ = vector_2d_angle(
+        ((int(hand_[0][0]) - int(hand_[10][0])), (int(hand_[0][1]) - int(hand_[10][1]))),
+        ((int(hand_[11][0]) - int(hand_[12][0])), (int(hand_[11][1]) - int(hand_[12][1])))
+    )
+    angle_list.append(angle_)
+    # ---------------------------- ring 无名指角度
```
Comment (WanliZhong) [inline models/handpose_estimation_mediapipe/demo.py:194]: Don't leave Chinese comments

File: models/handpose_estimation_mediapipe/demo.py [hunk 9]
```diff
@@ -146,6 +156,126 @@ def draw_lines(image, landmarks, is_draw_point=True, thickness=2):
...
+    angle_ = vector_2d_angle(
+        ((int(hand_[0][0]) - int(hand_[14][0])), (int(hand_[0][1]) - int(hand_[14][1]))),
+        ((int(hand_[15][0]) - int(hand_[16][0])), (int(hand_[15][1]) - int(hand_[16][1])))
+    )
+    angle_list.append(angle_)
+    # ---------------------------- pink 小拇指角度
```
Comment (WanliZhong) [inline models/handpose_estimation_mediapipe/demo.py:200]: Don't leave Chinese comments

File: models/handpose_estimation_mediapipe/demo.py [hunk 10]
```diff
@@ -214,6 +344,29 @@ def draw_lines(image, landmarks, is_draw_point=True, thickness=2):
...
+                        for i in range(21):
+                            x = landmarks[i][0] * frame.shape[1]
+                            y = landmarks[i][1] * frame.shape[0]
+                            hand_local.append((x, y))
+
+                        if hand_local:
+                            gesture = h_gesture(hand_local)
+                        if gesture is not None:
+                            tmp = dict[gesture]
+                            cv.putText(frame, gesture, (1000, 100), 0, 1.3, (0, 0, 255), 3)
```
Comment (WanliZhong) [inline models/handpose_estimation_mediapipe/demo.py:369]: The code about visualization needs to be placed in `visualize` function.

**Review by WanliZhong — COMMENTED** [review]
👍👍👍，Thanks for your contribution! Please, take a look at the comments below.

File: models/handpose_estimation_mediapipe/demo.py [hunk 11]
```diff
@@ -9,6 +9,9 @@
 sys.path.append('../palm_detection_mediapipe')
 from mp_palmdet import MPPalmDet
 
+dict = {"zero": 0, "one": 1, "two": 2, "three": 3, "four": 4, "five": 5, "six": 6, "seven": 7, "eight": 8, "nine": 9}
```
Comment (fengyuentau) [inline models/handpose_estimation_mediapipe/demo.py:12]: `dict` is a keyword in Python. Use another name.

File: models/handpose_estimation_mediapipe/demo.py [hunk 12]
```diff
@@ -17,6 +20,7 @@ def str2bool(v):
     else:
         raise NotImplementedError
 
+
```
Comment (fengyuentau) [inline models/handpose_estimation_mediapipe/demo.py:23]: unnecessary new line

File: models/handpose_estimation_mediapipe/demo.py (hunk 10 above)
Comment (fengyuentau) [reply models/handpose_estimation_mediapipe/demo.py:369]: One more suggestion here is name the function something like `gesture_classification` to make it more meaningful.

**Review by fengyuentau — COMMENTED** [review]
Thank you for contribution! I have some minor comments below.

**Comment by fengyuentau** [comment]
@yqxzy friendly reminder.

**Pull request closed without merging**